jetty-core/jetty-http2/jetty-http2-server/src/main/java/org/eclipse/jetty/http2/server/internal/HTTP2ServerConnection.java

```diff
@@ -223,83 +227,37 @@ public void push(HTTP2Stream stream, MetaData.Request request)
             LOG.debug("Processing push {} on {}", request, stream);
 
-        HttpChannel httpChannel = httpChannelFactory.newHttpChannel(this);
+        HttpChannel httpChannel = pollHttpChannel();
         HttpStreamOverHTTP2 httpStream = new HttpStreamOverHTTP2(this, httpChannel, stream);
         httpChannel.setHttpStream(httpStream);
         Runnable task = httpStream.onPushRequest(request);
         if (task != null)
             offerTask(task, true);
     }
 
-/*
-    // TODO: re-instate recycle channel functionality, but using Pool.
-    private final AutoLock lock = new AutoLock();
-    private final Queue<HttpChannelOverHTTP2> channels = new ArrayDeque<>();
-    private boolean recycleHttpChannels = true;
-
-    public boolean isRecycleHttpChannels()
-    {
-        return recycleHttpChannels;
-    }
-
-    public void setRecycleHttpChannels(boolean recycleHttpChannels)
+    private HttpChannel pollHttpChannel()
     {
-        this.recycleHttpChannels = recycleHttpChannels;
+        HttpChannel httpChannel = isRecycleHttpChannels() ? httpChannels.poll() : null;
+        if (httpChannel == null)
+            httpChannel = httpChannelFactory.newHttpChannel(this);
+        return httpChannel;
     }
 
-    private HttpChannelOverHTTP2 provideHttpChannel(Connector connector, IStream stream)
+    void offerHttpChannel(HttpChannel channel)
     {
-        HttpChannelOverHTTP2 channel = pollHttpChannel();
-        if (channel != null)
-        {
-            channel.getHttpTransport().setStream(stream);
-            if (LOG.isDebugEnabled())
-                LOG.debug("Recycling channel {} for {}", channel, this);
-        }
-        else
-        {
-            HttpTransportOverHTTP2 transport = new HttpTransportOverHTTP2(connector, this);
-            transport.setStream(stream);
-            channel = newServerHttpChannelOverHTTP2(connector, httpConfig, transport);
-            channel.setUseOutputDirectByteBuffers(isUseOutputDirectByteBuffers());
-            if (LOG.isDebugEnabled())
-                LOG.debug("Creating channel {} for {}", channel, this);
-        }
-        stream.setAttachment(channel);
-        return channel;
-    }
-
-    protected ServerHttpChannelOverHTTP2 newServerHttpChannelOverHTTP2(Connector connector, HttpConfiguration httpConfig, HttpTransportOverHTTP2 transport)
-    {
-        return new ServerHttpChannelOverHTTP2(connector, httpConfig, getEndPoint(), transport);
+        if (isRecycleHttpChannels())
+            httpChannels.offer(channel);
     }
 
-    private void offerHttpChannel(HttpChannelOverHTTP2 channel)
+    public boolean isRecycleHttpChannels()
     {
-        if (isRecycleHttpChannels())
-        {
-            try (AutoLock l = lock.lock())
-            {
-                channels.offer(channel);
-            }
-        }
+        return recycleHttpChannels;
     }
 
-    private HttpChannelOverHTTP2 pollHttpChannel()
+    public void setRecycleHttpChannels(boolean recycleHttpChannels)
```

Comment:
This config needs to be wired?

Reply:
Again, I took this from the commented out previous implementation.   I think it can be permanently on and not wired into config.   It is only there as an option in case we suspect a problem.

Reply:
But if it's not wired, it's useless as there is no way a user can enable it, and we can easily remove this recycling logic locally if we ever suspect it's causing problems and are working on reproducing the issue.
I think this boolean should just go and have the queue logic always on.

Reply:
@lorban I think it needs to be restored, so it will be easier to performance test the effect of recycling channels.
For a benchmark we can subclass `HTTP2ServerConnectionFactory.newConnection()` and explicitly call the boolean setter on the `HTTP2Connection`.